File: frontend/passes/linear_lower.go
package passes

import (
	"github.com/mattcarp12/maml/frontend/mir"
)

// LowerLinearTypes scrubs borrow-checker-specific instructions and lowers
// them into standard memory operations so the C++/LLVM backend doesn't
// have to reason about NLL lifetimes or capabilities.
func LowerLinearTypes(g *mir.Graph) {
	for _, block := range g.Blocks {
		var newStmts []mir.Instruction
		for _, inst := range block.Statements {
			switch i := inst.(type) {
			case *mir.BorrowInst:
				newStmts = append(newStmts, &mir.AddressOfInst{
					Dst: i.Dst,
					Src: i.Src,
				})

			case *mir.KeepAliveInst:
				continue

			default:
				newStmts = append(newStmts, inst)
			}
		}
		block.Statements = newStmts
	}
}

File: passes/src/linear_lower.cpp
#include "linear_lower.h"

namespace maml::passes {

void lowerLinearTypes(mir::Graph* g)
{
    if (!g)
        return;

    for (auto& [_, block] : g->blocks) {
        std::vector<mir::Instruction> newStmts;
        for (auto& inst : block->statements) {
            if (auto* borrow = std::get_if<mir::BorrowInst>(&inst)) {
                newStmts.push_back(mir::AddressOfInst { borrow->dst, borrow->src, borrow->pos });
            } else if (std::holds_alternative<mir::KeepAliveInst>(inst)) {
                continue; // Strip KeepAlive instructions[cite: 3]
            } else {
                newStmts.push_back(std::move(inst));
            }
        }
        block->statements = std::move(newStmts);
    }
}

} // namespace maml::passes